sbi/utils/posterior_ensemble.py

```diff
@@ -0,0 +1,446 @@
+                This can be helpful for e.g. sensitivity analysis, but increases memory
+                consumption.
+        """
+        self.potential_fn = EnsemblePotentialProvider(
+            self.posteriors, self._weights, self.prior, x
+        )
+        theta = ensure_theta_batched(torch.as_tensor(theta))
+
+        return self.potential_fn(
+            theta.to(self._device), track_gradients=track_gradients
+        )
+
+    def map(
+        self,
+        x: Optional[Tensor] = None,
+        num_iter: int = 1_000,
+        num_to_optimize: int = 100,
+        learning_rate: float = 0.01,
+        init_method: Union[str, Tensor] = "posterior",
+        num_init_samples: int = 1_000,
+        save_best_every: int = 10,
+        show_progress_bars: bool = False,
+        individually: bool = False,
+    ) -> Tensor:
+        r"""Returns the average maximum-a-posteriori estimate (MAP).
+
+        Computes MAP estimate across the whole ensemble or for each component
+        individually. All args and kwargs are passed directly through to
+        `gradient_ascent`.
+
+        The routine can be interrupted (individually) with [Ctrl-C], when the user sees
+        that the log-probability converges. The best estimate will be saved in `self.
+        posteriors[idx].map_`.
+
+        For more details of how the MAP estimate is obtained see `.map()` docstring of
+        self.posteriors[idx].
+
+        Args:
+            x: Observed data at which to evaluate the MAP.
+            num_iter: Number of optimization steps that the algorithm takes
+                to find the MAP.
+            num_to_optimize: From the drawn `num_init_samples`, use the
+                `num_to_optimize` with highest log-probability as the initial points
+            learning_rate: Learning rate of the optimizer.
+            init_method: How to select the starting parameters for the optimization. If
+                it is a string, it can be either [`posterior`, `prior`], which samples
+                the respective distribution `num_init_samples` times. If it is a
+                tensor, the tensor will be used as init locations.
+            num_init_samples: Draw this number of samples from the posterior and
+                evaluate the log-probability of all of them.
+                for the optimization.
+            save_best_every: The best log-probability is computed, saved in the
+                `map`-attribute, and printed every `save_best_every`-th iteration.
+                Computing the best log-probability creates a significant overhead
+                (thus, the default is `10`.)
+            show_progress_bars: Whether or not to show a progressbar for sampling from
+                the posterior.
+            individually: If true, returns log weights and MAPs individually.
+
+        Returns:
+            The ensemble MAP estimate or individual log_weigths and component MAP
+            estimate if individually == True.
+        """
+
```

Comment:
Here, you have to call the [_x_else_default_x](https://github.com/mackelab/sbi/blob/8657c55b39ec8f673a0eaa0cb2428b7f2ce5b14f/sbi/inference/posteriors/base_posterior.py#L134-L145) method. If `x=None`, default_x will be used.

Reply:
I thought I could just rely on the class method of the individual posteriors that gets called downstream anyway?